File: algoritmo.js
function vertice(nomeVertice, arestas) {
    return {
        "nome": nomeVertice,
        "arestas": arestas,
        "Conexao": function (vertice, valor) {
            let arestaLigacao = aresta(this, vertice, valor);
            this.arestas.push(arestaLigacao);
            vertice.arestas.push(arestaLigacao);
        },
        "verticesVizinhos": function () {
            let listaDeVerticesVizinhos = []
            for (let contagem = 0; contagem < this.arestas.length; contagem++) {
                if (this.arestas[contagem].verticeUm.nome == this.nome) {
                    listaDeVerticesVizinhos.push({
                        "vertice": this.arestas[contagem].verticeDois,
                        "peso": this.arestas[contagem].valor
                    })
                }
                else {
                    listaDeVerticesVizinhos.push({
                        "vertice": this.arestas[contagem].verticeUm,
                        "peso": this.arestas[contagem].valor
                    })
                }
            }
            return listaDeVerticesVizinhos;
        }
    };
};

function aresta(verticeUm, verticeDois, valor) {
    return {
        verticeUm, verticeDois, valor
    };
};

function visitante() {
    let visitados = []
    return {
        "visitados": visitados,
        "adicionar": function (verticeVisitado) {
            this.visitados.push(verticeVisitado)
        },
        "foiVisitado": function (verticeFoiVisitado) {
            for (let contagem = 0; contagem < this.visitados.length; contagem++) {
                if (this.visitados[contagem].nome == verticeFoiVisitado.nome) {
                    return true
                }
            }
            return false
        }
    }
}

function dijkstra(inicio, fim) {
    let distancias = {};
    let anterior = {};
    let naoVisitados = [];
    let visitador = visitante();
    distancias[inicio.nome] = 0;
    naoVisitados.push(inicio);
    while (naoVisitados.length > 0) {
        naoVisitados.sort((a, b) => distancias[a.nome] - distancias[b.nome]);
        let verticeAtual = naoVisitados.shift()
        if (verticeAtual.nome == fim.nome) {
            break;
        }
        visitador.adicionar(verticeAtual);
        let vizinhos = verticeAtual.verticesVizinhos();
        vizinhos.forEach(function (vizinho) {
            if (!visitador.foiVisitado(vizinho.vertice)) {
                let distanciaAtual = distancias[verticeAtual.nome] + vizinho.peso;
                if (distanciaAtual < (distancias[vizinho.vertice.nome] || Infinity)) {
                    distancias[vizinho.vertice.nome] = distanciaAtual;
                    anterior[vizinho.vertice.nome] = verticeAtual;
                    naoVisitados.push(vizinho.vertice);
                }
            }
        });
    }
    let caminho = [];
    let atual = fim;
    while (atual) {
        caminho.unshift(atual.nome);
        atual = anterior[atual.nome];
    }
    return {
        "caminho": caminho,
        "distancia": distancias[fim.nome]
    }
}

let verticeA = vertice("A", [])
let verticeB = vertice("B", [])
let verticeC = vertice("C", [])
let verticeD = vertice("D", [])
let verticeE = vertice("E", [])
let verticeF = vertice("F", [])
let verticeG = vertice("G", [])
let verticeH = vertice("H", [])
let verticeI = vertice("I", [])
let verticeJ = vertice("J", [])
let verticeK = vertice("K", [])
let verticeL = vertice("L", [])

verticeA.Conexao(verticeB, 2);
verticeA.Conexao(verticeD, 4);
verticeB.Conexao(verticeC, 3);
verticeB.Conexao(verticeL, 7);
verticeC.Conexao(verticeD, 5);
verticeD.Conexao(verticeE, 2);
verticeE.Conexao(verticeF, 5);
verticeE.Conexao(verticeI, 6);
verticeF.Conexao(verticeL, 4);
verticeF.Conexao(verticeG, 6);
verticeG.Conexao(verticeK, 3);
verticeG.Conexao(verticeH, 3);
verticeH.Conexao(verticeI, 4);
verticeI.Conexao(verticeJ, 10);
verticeJ.Conexao(verticeK, 8);
verticeK.Conexao(verticeL, 6);

let resultado = dijkstra(verticeA, verticeK);
console.log(resultado.caminho);

function medirTempoExecucao(funcao) {
    const inicio = performance.now();
    funcao(); 
    const fim = performance.now(); 
    const tempoExecucao = fim - inicio;
    console.log(`Tempo de execução: ${tempoExecucao.toFixed(2)} ms`);
    return tempoExecucao; 
}

const tempo = medirTempoExecucao(() => {
    for (let i = 0; i < 1000000; i++) {
        Math.sqrt(i);
    }
});

File: algoritmo.py
class Vertice:
    def __init__(self, nome):
        self.nome = nome
        self.arestas = []

    def conexao(self, vertice, valor):
        aresta_ligacao = Aresta(self, vertice, valor)
        self.arestas.append(aresta_ligacao)
        vertice.arestas.append(aresta_ligacao)

    def vertices_vizinhos(self):
        lista_vizinhos = []
        for aresta in self.arestas:
            if aresta.vertice_um.nome == self.nome:
                lista_vizinhos.append({"vertice": aresta.vertice_dois, "peso": aresta.valor})
            else:
                lista_vizinhos.append({"vertice": aresta.vertice_um, "peso": aresta.valor})
        return lista_vizinhos


class Aresta:
    def __init__(self, vertice_um, vertice_dois, valor):
        self.vertice_um = vertice_um
        self.vertice_dois = vertice_dois
        self.valor = valor


class Visitante:
    def __init__(self):
        self.visitados = []

    def adicionar(self, vertice):
        self.visitados.append(vertice)

    def foi_visitado(self, vertice):
        return vertice in self.visitados


def dijkstra(inicio, fim):
    distancias = {inicio.nome: 0}
    anterior = {}
    nao_visitados = [inicio]
    visitador = Visitante()

    while nao_visitados:
        nao_visitados.sort(key=lambda v: distancias.get(v.nome, float('inf')))
        vertice_atual = nao_visitados.pop(0)

        if vertice_atual.nome == fim.nome:
            break

        visitador.adicionar(vertice_atual)
        vizinhos = vertice_atual.vertices_vizinhos()

        for vizinho in vizinhos:
            vertice_vizinho = vizinho["vertice"]
            peso = vizinho["peso"]
            if not visitador.foi_visitado(vertice_vizinho):
                distancia_atual = distancias[vertice_atual.nome] + peso
                if distancia_atual < distancias.get(vertice_vizinho.nome, float('inf')):
                    distancias[vertice_vizinho.nome] = distancia_atual
                    anterior[vertice_vizinho.nome] = vertice_atual
                    nao_visitados.append(vertice_vizinho)

    caminho = []
    atual = fim
    while atual:
        caminho.insert(0, atual.nome)
        atual = anterior.get(atual.nome)

    return {"caminho": caminho, "distancia": distancias.get(fim.nome, float('inf'))}



verticeA = Vertice("A")
verticeB = Vertice("B")
verticeC = Vertice("C")
verticeD = Vertice("D")
verticeE = Vertice("E")
verticeF = Vertice("F")
verticeG = Vertice("G")
verticeH = Vertice("H")
verticeI = Vertice("I")
verticeJ = Vertice("J")
verticeK = Vertice("K")
verticeL = Vertice("L")


verticeA.conexao(verticeB, 2)
verticeA.conexao(verticeD, 4)
verticeB.conexao(verticeC, 3)
verticeB.conexao(verticeL, 7)
verticeC.conexao(verticeD, 5)
verticeD.conexao(verticeE, 2)
verticeE.conexao(verticeF, 5)
verticeE.conexao(verticeI, 6)
verticeF.conexao(verticeL, 4)
verticeF.conexao(verticeG, 6)
verticeG.conexao(verticeK, 3)
verticeG.conexao(verticeH, 3)
verticeH.conexao(verticeI, 4)
verticeI.conexao(verticeJ, 10)
verticeJ.conexao(verticeK, 8)
verticeK.conexao(verticeL, 6)

resultado = dijkstra(verticeA, verticeK)
print(resultado["caminho"])

import time

def medir_tempo_execucao(funcao):
    inicio = time.time()
    funcao()
    fim = time.time()
    tempo_execucao = (fim - inicio) * 1000
    print(f"Tempo de execução: {tempo_execucao:.2f} ms")
    return tempo_execucao


medir_tempo_execucao(lambda: [x**0.5 for x in range(1000000)])
